npk/src/manifest.rs

```diff
@@ -26,134 +25,200 @@ use std::{
+    /// String containing capability names to give to
+    /// new container
+    #[serde(default, with = "serde_caps")]
+    pub capabilities: Option<HashSet<Capability>>,
+    /// String containing group names to give to new container
+    pub suppl_groups: Option<Vec<String>>,
+    /// IO configuration
+    pub io: Option<Io>,
 }
 
-impl FromStr for Version {
-    type Err = semver::SemVerError;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Self::parse(s)
-    }
-}
+impl Manifest {
+    /// Manifest version supported by the runtime
+    pub const VERSION: Version = Version {
+        major: 0,
+        minor: 1,
+        patch: 0,
+        pre: vec![],
+        build: vec![],
+    };
 
-impl Default for Version {
-    fn default() -> Version {
-        Version(semver::Version::new(0, 0, 0))
+    pub fn from_reader<R: io::Read>(reader: R) -> Result<Self, Error> {
+        let manifest: Self = serde_yaml::from_reader(reader).map_err(Error::SerdeYaml)?;
+        manifest.verify()?;
+        Ok(manifest)
     }
-}
 
-/// Serde serialization for `Version`
-impl Serialize for Version {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_str(&self.0.to_string())
+    pub fn to_writer<W: io::Write>(&self, writer: W) -> Result<(), Error> {
+        serde_yaml::to_writer(writer, self).map_err(Error::SerdeYaml)
     }
-}
 
-/// Serde deserialization for `Version`
-impl<'de> Deserialize<'de> for Version {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        struct VersionVisitor;
+    fn verify(&self) -> Result<(), Error> {
+        // TODO: check for none on env, autostart, cgroups, seccomp
+        if self.init.is_none() && self.args.is_some() {
+            return Err(Error::Invalid(
+                "Arguments not allowed in resource container".to_string(),
+            ));
+        }
 
-        impl<'de> Visitor<'de> for VersionVisitor {
-            type Value = Version;
-            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> ::std::fmt::Result {
-                formatter.write_str("string v0.0.0")
-            }
+        // Check for invalid uid 0
+        if self.uid == 0 {
+            return Err(Error::Invalid("Invalid uid 0".to_string()));
+        }
 
-            fn visit_str<E>(self, str_data: &str) -> Result<Version, E>
-            where
-                E: serde::de::Error,
-            {
-                semver::Version::parse(str_data).map(Version).map_err(|_| {
-                    serde::de::Error::invalid_value(::serde::de::Unexpected::Str(str_data), &self)
-                })
+        // Check for null bytes in suppl groups. Rust Strings allow null bytes in Strings.
+        // For passing the group names to getgrnam they need to be C string compliant.
+        if let Some(suppl_groups) = self.suppl_groups.as_ref() {
```

Comment:
Would it make sense to replace these Strings in the manifest with CStrings right from the start?

Reply:
Interesting.

A `CString` is basically a `&[u8]`. In theory we can allow in the manifest everything that does not contains a null byte. Currently we force to have valid utf8 in the manifest which also doesn't allow any null byte and is more restrictive than the C representation.

Is there a use case where a non utf8 string shall be used as `init`, in `env`, `suppl_groups` etc?

Reply:
I created #367